kernel/src/protocols/attestation.rs

```diff
@@ -0,0 +1,356 @@
+        let size = self.manifest_size as usize;
+
+        Ok((gpa, size))
+    }
+
+    /// Returns the guid
+    /// Checks if gpa is page aligned, valid  and does not cross page boundary
+    pub fn get_report_gpa_and_size(&self) -> Result<(PhysAddr, usize), SvsmReqError> {
+        let gpa = PhysAddr::from(self.report_gpa);
+        if !gpa.is_aligned(8) || !valid_phys_address(gpa) || gpa.crosses_page(8) {
+            return Err(SvsmReqError::invalid_parameter());
+        }
+
+        // Won't panic on amd64 as usize > u32 always
+        let size = self.report_size as usize;
+
+        Ok((gpa, size))
+    }
+
+    /// Returns the certificate GPA and size
+    /// Checks if gpa is page aligned, valid  and does not cross page boundary
+    /// pub fn get_report_gpa_and_size(&self) -> Result<(PhysAddr, usize), SvsmReqError> {
+    /// Currently not implemented supported as no service attestation currently supports
+    /// returning certificates. Implement when needed by copying get_report_gpa_and_size()
+    /// and replacing report's gpa and size  with certificate's.
+    /// }
+
+    pub fn get_manifest_version(&self) -> u32 {
+        self.manifest_ver
+    }
+
+    fn is_manifest_version_valid(&self) -> bool {
+        //Currently only manifest version 0 is supported
+        self.manifest_ver == 0
+    }
+    pub fn get_guid(&self) -> u128 {
+        u128::from_le_bytes(self.guid)
+    }
+}
+
+#[cfg(all(feature = "vtpm", not(test)))]
+fn get_attestation_report(nonce: &[u8]) -> Result<Vec<u8>, SvsmReqError> {
+    //Construct attestation request message to send to SNP
+    let mut report_req = Vec::<u8>::with_capacity(size_of::<SnpReportResponse>());
+    let mut buf = Vec::<u8>::with_capacity(USER_DATA_SIZE);
+    buf.fill(0);
+
+    if nonce.len() > USER_DATA_SIZE {
+        // If the nonce is greater than the user data size, return an error as something is wrong.
+        return Err(SvsmReqError::invalid_parameter());
+    }
+    // Copy user attestation request nonce to buffer
+    buf.extend_from_slice(nonce);
+    report_req.extend_from_slice(&buf[..nonce.len()]);
+
+    // Set request VMPL to 0
+    report_req.extend_from_slice(&0_u32.to_le_bytes());
+
+    // Set reserved bytes to zeros
+    report_req.extend_from_slice(&[0; 28]);
+
+    // Make sure buffer is big enough to hold the report
+    report_req.resize(size_of::<SnpReportResponse>(), 0);
+
+    //send request to snp
+    let response_size = get_regular_report(report_req.as_mut_slice()).unwrap();
+
+    // Per Table 24 of "SEV Secure Nested Paging Firmware ABI Specification, Revision 1.56",
+    // attestation report starts  at byte offset 0x20
+    if response_size < 0x20 {
+        // If the response size is less than 0x20, return an error as something is wrong with
+        //  the report or SNP
+        return Err(SvsmReqError::unsupported_protocol());
+    }
+    report_req.drain(0..0x20);
+
+    Ok(report_req)
```

Comment:
validate() is called inside get_report() which is called by get_regular_report().

Reply:
> I wanted to ref link the to the function in the comment but can't run "make doc" any more. Something added recently broke it.

I just tried on the `main` branch and it works, so it may be related to the changes we did to our submodule.  See https://github.com/coconut-svsm/svsm/issues/521

Reply:
Thanks that worked